Spec feature profiles

Kegsay opened the pull request:
The intent of this PR is to document:
 - What are feature profiles
 - What are clients
 - Which features go to which clients and why

This PR is not intended to:
 - Delve into the docs for any particular module

These feature profiles are based off https://github.com/matrix-org/matrix-doc/pull/52#issuecomment-141699759

The subsections for each client are somewhat sparse, I was thinking that we could add some justifications to the selected features there, but wanted to get the ball rolling on some feedback first.

As a side note, I don't feel that `history_visibility` should actually be a module - it's surely room-ACL stuff which all clients will be bound by regardless of what they implement. It should be a YAML schema thing instead. If people agree, I will move it over in another PR.

illicitonion commented on specification/0-feature_profiles.rst:
```diff
@@ -1,3 +1,83 @@
+============================ ===== =========== ======== ========= ===== =====
+  Module / Profile            Web   Embed-Web   Mobile   Desktop   CLI   IoT
+============================ ===== =========== ======== ========= ===== =====
+ `End-to-End Encryption`_     YES                YES       YES     YES
```
Are we really mandating that all web/mobile/desktop clients *must* support e2e?

The e2e section reads "Matrix optionally supports end-to-end encryption".

If we are mandating it must be supported, are we also going to mandate a base level of algorithms which *must* be supported?

illicitonion commented on specification/0-feature_profiles.rst:
```diff
@@ -1,3 +1,83 @@
+Desktop (``Desktop``)
+~~~~~~~~~~~~~~~~~~~~~
+
+This is a native application which can run in its own environment outside a
```
This isn't in any way distinguished from CLI

Kegsay replied:
Done.

ara4n commented:
Generally looks good to me.

Given E2E isn't finished yet suggest we leave it as optional across the board. 

I think we should keep history ACLs as their own module (as they are a fairly standalone component of the spec, and we already have different proposals flying around) but make support for understanding ACL events mandatory across all human facing clients. In practice this will mean doing the "do you want to share your history with the new user?" prompt and setting the appropriate ACL state event.

We probably need client profiles for embedded mobile clients and embedded desktop clients - equivalent to embedded web but for native apps (eg adding Matrix chat support to a native wordprocessor, or PTT audio to a native game, etc).

This is currently all very client specific - i wonder if we should be defining server profiles (eg "crappy HS with no support for E2E, or persistence, or even federation") - @NegativeMjark, what did you have in mind here? Personally i'd leave it as client behaviour profiles for now.

We probably also need feature profiles for ASes at some point too. "This is what i means to be a bridge" etc.

Kegsay commented:
> Given E2E isn't finished yet suggest we leave it as optional across the board.

ACK. Will do.

> I think we should keep history ACLs as their own module (as they are a fairly standalone component of the spec, and we already have different proposals flying around) but make support for understanding ACL events mandatory across all human facing clients. In practice this will mean doing the "do you want to share your history with the new user?" prompt and setting the appropriate ACL state event.

I see, but this feels wrong to me. This kind of logic can be extended to almost anything (setting aliases, setting join rules, setting power levels, etc) - I feel the "core" (non-modules) spec should consist of:
 - Registration / Login
 - Room creation
 - Memberships (invite/join/kick/ban/knock/etc)
 - ACLs (`join_rules`, `power_levels`, `history_visibility`, etc)
 - Pagination
 - Syncing (initial and subsequent)
 - Sending events to a room (non-specific about types)

I find it hard to justify what makes `history_visibility` special.

> i wonder if we should be defining server profiles (eg "crappy HS with no support for E2E, or persistence, or even federation")

I don't think fragmenting servers functionally is a good way to do it because you need to then do lots of capability negotation. Fragmenting them based on what it can speak to is better imo, e.g:
 - Client module (Implements C-S API to communicate to clients)
 - Federation module (to communicate with other servers)
 - AS module (to communicate with ASes)
 - Push module (to communicate with pushers)

Push may be slightly too much since it does start impacting functionality (no push notifications) but the rest I feel are good boundaries in that they are relatively self-contained modules.

> We probably also need feature profiles for ASes at some point too. "This is what i means to be a bridge" etc.

This sounds reasonable: we want to make sure bridges are setting those $to-be-confirmed state events to prevent routing loops after all.

ara4n commented:
>     I think we should keep history ACLs as their own module (as they
>     are a fairly standalone component of the spec, and we already have
>     different proposals flying around) but make support for
>     understanding ACL events mandatory across all human facing
>     clients. In practice this will mean doing the "do you want to
>     share your history with the new user?" prompt and setting the
>     appropriate ACL state event.
>
> I see, but this feels wrong to me. This kind of logic can be extended 
> to almost anything (setting aliases, setting join rules, setting power 
> levels, etc) - I feel the "core" (non-modules) spec should consist of:
>
>   * Registration / Login
>   * Room creation
>   * Memberships (invite/join/kick/ban/knock/etc)
>   * ACLs (|join_rules|, |power_levels|, |history_visibility|, etc)
>   * Pagination
>   * Syncing (initial and subsequent)
>   * Sending events to a room (non-specific about types)
>

The rationale here is that:
  * You can use Matrix fine without any ACLs (as we did for a year or so)
  * ACLs are particularly fiddly to define correctly, and we actively 
expect the spec to iterate several times until we get it right - i.e. we 
have both SPEC-134 and SPEC-210; making the modular makes it easier to 
swap out the spec.

So whilst we expect to make them compulsory across the board, 
ringfencing the spec from the rest of the 'core' makes it easier to 
evolve and swap out independently.

Kegsay commented:
Hmm okay, will give it a go.

illicitonion commented on specification/0-feature_profiles.rst:
```diff
@@ -1,3 +1,85 @@
+============================ ===== =========== ======== ========= ===== =====
+  Module / Profile            Web   Embed-Web   Mobile   Desktop   CLI   IoT
+============================ ===== =========== ======== ========= ===== =====
+ `End-to-End Encryption`_                                             
```
Why do we have something here which isn't required by any implementation?

Kegsay replied:
Because it is still a module.

illicitonion commented:
@ara4n: I don't understand how we both can view things is independent and live-without-able, and take an attitude that we need a large set of core functionality and that extensions will cause intolerable fragmentation.

ara4n commented:
As per RL, this PR *avoids* fragmentation by specifying which modules are *mandatory* for classes of client.  Things like the history ACLs which are mandatory for all clients are just split out as a separate module for maintainability and because we expect them to evolve.  Sure, evolving will cause some fragmentation, but we can always try to do so whilst retaining backwards compatibility - and Matrix has to keep evolving as a spec otherwise it will just paint itself into a corner.

illicitonion commented:
LGTM

ara4n commented:
<Arathorn> kegan: on PR 65, there's already pending feedback from me
 <Arathorn> plus surely CLI shouldn't require content repo
 <Arathorn> and embed-web needs history visibility
 <Arathorn> i'd put e2e at the bottom of the list
 <Arathorn> in the table at http://matrix.org/speculator/spec/65#id41
 <Arathorn> and we need Embedded Mobile
 <Arathorn> and Embedded Desktop
 <Arathorn> as things
 <Arathorn> i also query if we should mandate IM for Embedded clients
 <Arathorn> as one could easily just use matrix for (say) VoIP and VoIP alone
 <Arathorn> in an embedded context
 <Arathorn> so perhaps we just turn the 3 embeddeds into a single Embed
 <Arathorn> and spell out that really eeeverything is optional.
 <kegan> can you dump ^ on the PR please? 0:)
 <Arathorn> this could also let us kill the IOT profile too, and declare that IOT is "Embedded" (embedded in a device, rather than embedded in an app, but the features still apply)

ara4n commented:
LeoNerd: I'd be tempted to put "(opt)" in all the other cells, or something.. some way to indicate "hey, we don't forbid anyone implementing another feature outside their set"; I.e. that it isn't "No"
^ agreed.
Also, we might want to give some examples for the different clients.  E.g. give Vector an example of Web, Weechat CLI, Drone or h4x.no's blog as Embedded, etc.

Kegsay commented:
I think I've addressed all the comments. I would prefer to give generic examples rather than specific instances of real-world clients to avoid any favouritism conflicts which may arise (e.g. "why isn't my app mentioned in the spec" :cry: )

ara4n commented:
On 30/09/2015 16:50, Kegsay wrote:
> I would prefer to give
> generic examples rather than specific instances of real-world clients
> to avoid any favouritism conflicts which may arise (e.g. "why isn't
> my app mentioned in the spec" :cry: )

okay, fair enough.

ara4n commented:
As per conversation at lunch, my final request would be to move this section to be immediately before Modules - either as a top level section (to give it maximum importance), or possibly as a prelude to the list of Modules.

ara4n commented on specification/0-feature_profiles.rst:
```diff
@@ -1,3 +1,93 @@
+ `Typing Notifications`_      Required   Required   Required   Required   Optional
+ `VoIP`_                      Required   Required   Required   Optional   Optional
+ `Content Repository`_        Required   Required   Required   Optional   Optional
+ `History Visibility`_        Required   Required   Required   Required   Optional
```
Awareness of history visibility is surely required even on embedded clients.

Kegsay replied:
Not so sure. I'm basing this from your earlier comment that:

> understanding ACL events [should be] mandatory across all human facing clients. In practice this will mean doing the "do you want to share your history with the new user?" prompt and setting the appropriate ACL state event.

For embedded web stuff sure, since it's human facing. For my toaster... Really? Maybe we should have Embedded Apps and Embedded Devices?

ara4n replied:
oh, hm, you might be right. I guess we need to make it clear precisely 
what the scope of the history visibility module is. Perhaps we should 
name it "managing history visibility" or something.

On 30/09/2015 17:48, Kegsay wrote:
> In specification/0-feature_profiles.rst
> <https://github.com/matrix-org/matrix-doc/pull/65#discussion_r40818937>:
>
>
>> +in order for it to be classified as that profile. + +Summary
>> +------- + +============================ ========== ==========
>> ========== ========== ========== +  Module / Profile
>> Web       Mobile    Desktop       CLI     Embedded
>> +============================ ========== ========== ==========
>> ========== ========== + `Instant Messaging`_         Required
>> Required   Required   Required   Optional + `Presence`_
>> Required   Required   Required   Required   Optional + `Push
>> Notifications`_        Optional   Required   Optional   Optional
>> Optional + `Receipts`_                  Required   Required
>> Required   Required   Optional + `Typing Notifications`_
>> Required   Required   Required   Required   Optional + `VoIP`_
>> Required   Required   Required   Optional   Optional + `Content
>> Repository`_        Required   Required   Required   Optional
>> Optional + `History Visibility`_        Required   Required
>> Required   Required   Optional
>
> Not so sure. I'm basing this from your earlier comment that:
>
> understanding ACL events [should be] mandatory across all human
> facing clients. In practice this will mean doing the "do you want to
> share your history with the new user?" prompt and setting the
> appropriate ACL state event.
>
> For embedded web stuff sure, since it's human facing. For my
> toaster... Really? Maybe we should have Embedded Apps and Embedded
> Devices?
>
> — Reply to this email directly or view it on GitHub
> <https://github.com/matrix-org/matrix-doc/pull/65/files#r40818937>.
>

Kegsay replied:
Renamed. Anything else?

Kegsay commented:
> As per conversation at lunch, my final request would be to move this section to be immediately before Modules - either as a top level section (to give it maximum importance), or possibly as a prelude to the list of Modules.

Already done: https://github.com/matrix-org/matrix-doc/commit/be9402b66fbb49c4c23c57e7aec93d1f1efc03f3#diff-c55ddd162b33bb6296b321642c27085aR9

Kegsay commented:
@ara4n anything else to do here?

ara4n commented:
lgtm

> On 1 Oct 2015, at 11:20, Kegsay <notifications@github.com> wrote:
> 
> @ara4n anything else to do here?
> 
> —
> Reply to this email directly or view it on GitHub.
>
